src/Parser/TypeParser.php

```diff
@@ -208,4 +219,63 @@ private function tryParseArray(TokenIterator $tokens, Ast\Type\TypeNode $type):
 		return $type;
 	}
 
+
+	private function parseArrayShape(TokenIterator $tokens, Ast\Type\TypeNode $type): Ast\Type\TypeNode
+	{
+		$tokens->consumeTokenType(Lexer::TOKEN_OPEN_CURLY_BRACKET);
+		$items = [$this->parseArrayShapeItem($tokens)];
+
+		while (!$tokens->tryConsumeTokenType(Lexer::TOKEN_CLOSE_CURLY_BRACKET)) {
+			$tokens->consumeTokenType(Lexer::TOKEN_COMMA);
+			$items[] = $this->parseArrayShapeItem($tokens);
+		}
+
+		return new Ast\Type\ArrayShapeNode($items);
+	}
+
+
+	private function parseArrayShapeItem(TokenIterator $tokens): Ast\Type\ArrayItemNode
+	{
+		try {
+			$tokens->pushSavePoint();
+			$key = $this->parseArrayShapeKey($tokens);
+			$optional = $tokens->tryConsumeTokenType(Lexer::TOKEN_NULLABLE);
+			$tokens->consumeTokenType(Lexer::TOKEN_COLON);
+			$value = $this->parse($tokens);
+			$tokens->dropSavePoint();
+
+			return new Ast\Type\ArrayItemNode($key, $optional, $value);
+		} catch (\PHPStan\PhpDocParser\Parser\ParserException $e) {
+			$tokens->rollback();
+			$value = $this->parse($tokens);
+
+			return new Ast\Type\ArrayItemNode(null, $optional, $value);
+		}
+	}
+
+	/**
+	 * @return ConstExprStringNode|ConstExprIntegerNode|IdentifierTypeNode
+	 */
+	private function parseArrayShapeKey(TokenIterator $tokens)
```

Comment:
missing native type `Ast\Type\Node` (or whatever is the proper name)

Reply:
Are you refering to the return type ? What would you suggest ?

Reply:
Ok, got it :)

Reply:
yes, imho the only common is `\PHPStan\PhpDocParser\Ast\Node`